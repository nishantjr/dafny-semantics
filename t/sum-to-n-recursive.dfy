method Verify(n : int) returns (r : int)
  requires n >= 0
  ensures  r == n*(n + 1) / 2
{
  if (n == 0) {
    r := n;
  }
  if (n > 0) {
    var s : int;
    s := Verify(n - 1);
    r := n + s;
  }
}

method Main()
  returns (r: int)
  requires true
  ensures  true
{
  assert(Verify(0)  == 0);
  assert(Verify(3) == 6);
}
