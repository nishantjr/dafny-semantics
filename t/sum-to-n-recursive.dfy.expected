    #Not ( {
      _0
    #Equals
      0
    } )
  #And
    #Not ( {
      _1
    #Equals
      ( _0 -Int 1 ) *Int ( _0 -Int 1 +Int 1 ) /Int 2
    } )
  #And
    {
      Result
    #Equals
      <generatedTop>
        <k>
          .
        </k>
        <methods>
          Main |-> method Main ( .ArgDecls ) returns ( r : int , .ArgDecls ) requires true ensures true { assert Verify ( 0 , .ResultExps ) == 0 ;  assert Verify ( 3 , .ResultExps ) == 6 ;  .Statements }
          Verify |-> method Verify ( n : int , .ArgDecls ) returns ( r : int , .ArgDecls ) requires n >= 0 ensures r == n * ( n + 1 ) / 2 { if ( n == 0 ) { r := n ;  .Statements }  if ( n > 0 ) { var s : int ;  s := Verify ( n - 1 , .ResultExps ) ;  r := n + s ;  .Statements }  .Statements }
        </methods>
        <stack>
          .List
        </stack>
        <store>
          3 |-> ( _0 -Int 1 , const )
          4 |-> ( _1:Int , var )
        </store>
        <env>
          n |-> 3
          r |-> 4
        </env>
        <nextLoc>
          5
        </nextLoc>
      </generatedTop>
    }
  #And
    {
      true
    #Equals
      _0 -Int 1 >=Int 0
    }
  #And
    {
      true
    #Equals
      _0 >=Int 0
    }
  #And
    {
      true
    #Equals
      _0 >Int 0
    }
#Or
    #Not ( {
      _0
    #Equals
      0
    } )
  #And
    {
      Result
    #Equals
      <generatedTop>
        <k>
          .
        </k>
        <methods>
          Main |-> method Main ( .ArgDecls ) returns ( r : int , .ArgDecls ) requires true ensures true { assert Verify ( 0 , .ResultExps ) == 0 ;  assert Verify ( 3 , .ResultExps ) == 6 ;  .Statements }
          Verify |-> method Verify ( n : int , .ArgDecls ) returns ( r : int , .ArgDecls ) requires n >= 0 ensures r == n * ( n + 1 ) / 2 { if ( n == 0 ) { r := n ;  .Statements }  if ( n > 0 ) { var s : int ;  s := Verify ( n - 1 , .ResultExps ) ;  r := n + s ;  .Statements }  .Statements }
        </methods>
        <stack>
          .List
        </stack>
        <store>
          0 |-> ( _0:Int , var )
          1 |-> ( _0 +Int ( _0 -Int 1 ) *Int ( _0 -Int 1 +Int 1 ) /Int 2 , var )
          2 |-> ( ( _0 -Int 1 ) *Int ( _0 -Int 1 +Int 1 ) /Int 2 , var )
        </store>
        <env>
          n |-> 0
          r |-> 1
          s |-> 2
        </env>
        <nextLoc>
          5
        </nextLoc>
      </generatedTop>
    }
  #And
    {
      _0 +Int ( _0 -Int 1 ) *Int ( _0 -Int 1 +Int 1 ) /Int 2
    #Equals
      _0 *Int ( _0 +Int 1 ) /Int 2
    }
  #And
    {
      true
    #Equals
      _0 -Int 1 >=Int 0
    }
  #And
    {
      true
    #Equals
      _0 >=Int 0
    }
  #And
    {
      true
    #Equals
      _0 >Int 0
    }
#Or
    {
      Result
    #Equals
      <generatedTop>
        <k>
          .
        </k>
        <methods>
          Main |-> method Main ( .ArgDecls ) returns ( r : int , .ArgDecls ) requires true ensures true { assert Verify ( 0 , .ResultExps ) == 0 ;  assert Verify ( 3 , .ResultExps ) == 6 ;  .Statements }
          Verify |-> method Verify ( n : int , .ArgDecls ) returns ( r : int , .ArgDecls ) requires n >= 0 ensures r == n * ( n + 1 ) / 2 { if ( n == 0 ) { r := n ;  .Statements }  if ( n > 0 ) { var s : int ;  s := Verify ( n - 1 , .ResultExps ) ;  r := n + s ;  .Statements }  .Statements }
        </methods>
        <stack>
          .List
        </stack>
        <store>
          0 |-> ( _0:Int , var )
          1 |-> ( _00:Int , var )
        </store>
        <env>
          n |-> 0
          r |-> 1
        </env>
        <nextLoc>
          2
        </nextLoc>
      </generatedTop>
    }
  #And
    {
      false
    #Equals
      _0 >=Int 0
    }
#Or
  {
    Result
  #Equals
    <generatedTop>
      <k>
        .
      </k>
      <methods>
        Main |-> method Main ( .ArgDecls ) returns ( r : int , .ArgDecls ) requires true ensures true { assert Verify ( 0 , .ResultExps ) == 0 ;  assert Verify ( 3 , .ResultExps ) == 6 ;  .Statements }
        Verify |-> method Verify ( n : int , .ArgDecls ) returns ( r : int , .ArgDecls ) requires n >= 0 ensures r == n * ( n + 1 ) / 2 { if ( n == 0 ) { r := n ;  .Statements }  if ( n > 0 ) { var s : int ;  s := Verify ( n - 1 , .ResultExps ) ;  r := n + s ;  .Statements }  .Statements }
      </methods>
      <stack>
        .List
      </stack>
      <store>
        0 |-> ( 0 , var )
        1 |-> ( 0 , var )
      </store>
      <env>
        n |-> 0
        r |-> 1
      </env>
      <nextLoc>
        2
      </nextLoc>
    </generatedTop>
  }
